src/main/scala/lms/core/stub.scala

```diff
@@ -926,7 +932,7 @@ trait Base extends EmbeddedControls with OverloadHack with lms.util.ClosureCompa
     Adapter.g.reflectWrite("P",Unwrap(x))(Adapter.CTRL)
 
   def printf(f: String, x: Rep[Any]*): Unit = {
-    Adapter.g.reflectWrite("printf",Backend.Const(f)::x.map(Unwrap).toList:_*)(Adapter.CTRL)
+    Adapter.g.reflectEffect("printf", Backend.Const(f)::x.map(Unwrap).toList: _*)()(Adapter.CTRL)
```

Comment:
this change is useless, but I am pinning it here because I want to say that we need to enforce a type error if users try to printf an array of char

Reply:
f.split("%").map(_,charAt(0)) zip(x.map(typemap)) foreach {
  case ('d', m) => assert(m == manifest[Int])
  ...
}

or

-- Parse the format string (list of characters) into an AST.
-- Example: "%d,%s"  →   (FInt (FOther ',' (FString FEnd)))
format : List Char -> Format
format ('%' :: 'd' :: cs ) = FInt ( format cs )
format ('%' :: 's' :: cs ) = FString ( format cs )
format ( c :: cs )         = FOther c ( format cs )
format []                  = FEnd

https://gist.github.com/chrisdone/672efcd784528b7d0b7e17ad9c115292

Reply:
this code needs to be fixed with format string checks. :) 

not intending to push bad code in the repo :)